### DISCUSSION TITLE
Remove widgets from screenshots/movies

### DISCUSSION THREAD

**Discussion by adedner** [pyvista/pyvista#4549] [Ideas]
Problem
--------------
In some cases having the widgets (e.g. sliders or planes) in the output is not desirable. It is possible to use the `Off/On` methods on each widget (not documented) but having a more user friendly approach might be a _nice to have_?

Suggestions
------------------
One option would be a simple `diable_widgets` and `enable_widgets` method on the plotter.
Alternative approach with more flexibility would be to have an extra argument when adding the widgets to specify if that widget should be included in any output (default would be to include to maintain current behavior).

Current situation
--------------------------
Example with current methods:
```
# preamble
import pyvista as pv
pv.set_plot_theme("document")
from pyvista import examples
vol = examples.download_brain()
plotter = pv.Plotter()
# this is needed to get a good 'origin' for the plane widget
plotter.add_mesh(vol.outline_corners(), color='k') 
```
Adding a plane widget to show a slice (which is to be turned off for a screenshot) and a second plane widget without controls to be added only to the screenshot to outline the plane:
```
# add a slice with plane widget
def slice(normal,origin):
    slices = vol.slice(normal=normal, origin=origin)
    plotter.add_mesh(slices, name="slice")    
plane = plotter.add_plane_widget(slice)
# use second widget to have the clip plane visible in screenshot without the widgets
# this is just to see if that is possible
plane2 = plotter.add_plane_widget(lambda *args:[], 
                                     origin_translation=False, normal_rotation=False, implicit=True)
plane2.Off()
```
show and make a screenshot using `Off/On` methods:
```
plotter.show()
# now set a specific normal (and if required origin) and make a screenshot without the widget
plane.SetNormal([-1,1,1])
slice(plane.GetNormal(),plane.GetOrigin())
plane.Off()
plane2.On()
plane2.SetNormal( plane.GetNormal() )
plane2.SetOrigin( plane.GetOrigin() )
plotter.render()
plotter.screenshot("test.png", return_img=False)
plane2.Off()
plane.On()
plotter.render()
```

Suggested replacement
-----------------------------------
```
# add a slice with plane widget
def slice(normal,origin):
    slices = vol.slice(normal=normal, origin=origin)
    plotter.add_mesh(slices, name="slice")    
plane = plotter.add_plane_widget(slice, output=False)
plane2 = plotter.add_plane_widget(lambda *args:[],  output=True,
                                     origin_translation=False, normal_rotation=False, implicit=True)
plane2.Off()
plotter.show()
# now set a specific normal (and if required origin) and make a screenshot without the widget
plane.SetNormal([-1,1,1])
slice(plane.GetNormal(),plane.GetOrigin())
plane2.SetNormal( plane.GetNormal() )
plane2.SetOrigin( plane.GetOrigin() )
plotter.screenshot("test.png", return_img=False)
```
Probably the `screenshot` method would have to store the current `On/Off` state of each widget change it to the desired `output` state and restore the old settings after output. This also requires two additional rendering steps I guess so perhaps too invasive/inefficient.

**Comment by edabor** [comment]
@adedner I started looking at your idea. For what is worth, I found that the key press event `i` hides/show all widgets.
https://vtk.org/doc/nightly/html/classvtk3DWidget.html

**Reply by banesullivan** [reply]
Nice find @beroda! We can emulate that keypress in a routine to save screenshots like:

```py
import pyvista as pv
from pyvista import examples

mesh = examples.download_knee_full()

def screenshot_no_widgets(plotter, *args, **kwargs):
    plotter.iren._simulate_keypress('i')  # hide widgets
    plotter.render()  # make sure to render after hiding
    plotter.screenshot(*args, **kwargs)
    plotter.iren._simulate_keypress('i')  # show widgets
    plotter.render()

p = pv.Plotter()
p.add_mesh_threshold(mesh)
p.show(auto_close=False)

screenshot_no_widgets(p, 'test.png')
```

![test](https://github.com/pyvista/pyvista/assets/22067021/e2c32795-715a-4dad-9f8b-0fb5bd176403)

**Reply by banesullivan** [reply]
This will also hide axes widgets which may not be desired

**Reply by banesullivan** [reply]
Or maybe not... it seems the `i` key may only hide the last added widget

**Reply by edabor** [reply]
I can confirm that key press `i` hides the **first** widget only when there are multiple widgets.

I found some interesting stuff:
- after pressing `i`, some user interaction is needed for the widget to effectively disappear. Do you have the same on you side ?
- when using `trame`, the key press event `i` (and many others natively implemented in vtk) are ignored. Do you confirm ?

FYI, it seems that the code you provided does not work on my machine:
![image](https://github.com/pyvista/pyvista/assets/117822458/1a9a73dc-e266-45a4-a9c3-5adc8d2a0101)


Before the error is raised, this warning is appearing:
![image](https://github.com/pyvista/pyvista/assets/117822458/060e7c6a-4aa1-4e7c-8e74-042609619fb9)
I tried all solutions to quit the render window (keys `e`, `q`, close the GUI) without any success.

Sorry for the multiple stuffs, if you want I can open another discussion/issue to further investigate :smile:

**Reply by adedner** [reply]
I tried the `i` and only got it to work with the `BackgroundPlotter` but then also observed that only one widget disappears. Other key press events also don't work in inline plotting, i.e., `w` for switching to wireframe. That seems to work with some other backends (e.g. panel).

For me the above code works fine and when changing it to
```
p = pv.Plotter()
p.add_mesh_threshold(m)
p.add_mesh_clip_plane(mesh)
```
only one widget dissapears - strangely enough it's the slider that is not shown, i.e., the first I added.
